runtime/databricks/automl_runtime/hyperopt/early_stop.py

```diff
@@ -0,0 +1,71 @@
+#
+# Copyright (C) 2021 Databricks, Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+import logging
+
+from hyperopt import STATUS_OK
+
+_logger = logging.getLogger(__name__)
+_logger_hyperopt = logging.getLogger("hyperopt-spark")
+_logger_hyperopt.setLevel(logging.ERROR)
+
+# minimum number of trials to run before early stopping is considered
+NO_EARLY_STOP_THRESHOLD = 40
+# hyperopt search will stop early if the loss doesn't improve after this number of iterations
+NO_PROGRESS_STOP_THRESHOLD = 20
+
+
+def early_stop_fn(trials, best_loss=None, no_progress_iters=0, completed_trial_ids=set()):
+    """
+    The function checks if hyperopt should stop searching given results from the runs.
```

Comment:
Let's link to the open-source hyperopt repo: https://github.com/hyperopt/hyperopt/blob/master/hyperopt/fmin.py#L487

(since the function signature is kind of confusing and worth explaining more)

Reply:
Oh I thought it's not in open source yet. Thanks for pointing it out!